<launch>
  <node pkg="jsk_pcl_ros" type="sample_simulate_tabletop_cloud.py"
        name="sample_simulate_tabletop_cloud" />
  <node pkg="jsk_pcl_ros" type="interactive_cuboid_likelihood"
        name="interactive_cuboid_likelihood" output="log"
        >
    <remap from="~input" to="plane_supported_cuboid_estimator/output/candidate_cloud" />
    <param name="sensor_frame" value="sensor_frame" />
  </node>
  <node pkg="jsk_interactive_marker" type="marker_6dof" name="sensor_frame">
    <rosparam>
      publish_tf: true
      tf_frame: sensor_frame
      frame_id: odom
      object_x: 0.1
      object_y: 0.1
      object_z: 0.1
      initial_z: 2.0
      initial_x: -0.5
      initial_orientation: [0.641, 0.759, 0.052, 0.102]
    </rosparam>
  </node>
  <node pkg="jsk_pcl_ros" type="plane_supported_cuboid_estimator"
        name="plane_supported_cuboid_estimator" output="log"
        >
    <param name="sensor_frame" value="sensor_frame" />
    <remap from="~input" to="sample_simulate_tabletop_cloud/output"/>
    <remap from="~input/polygon" to="sample_simulate_tabletop_cloud/output/polygon" />
    <remap from="~input/coefficients" to="sample_simulate_tabletop_cloud/output/coef" />
    <rosparam>
      init_local_position_z_min: 0.05
      init_local_position_z_max: 0.25
      step_x_variance: 0.0001
      step_y_variance: 0.002
      step_z_variance: 0.002
      step_roll_variance: 0.000001
      step_pitch_variance: 0.000001
      step_yaw_variance: 0.001
      step_dx_variance: 0.0
      step_dy_variance: 0.0
      step_dz_variance: 0.0
      init_dx_variance: 0.0
      init_dy_variance: 0.0
      init_dz_variance: 0.0
      init_dx_mean: 0.1
      init_dy_mean: 0.1
      init_dz_mean: 0.1
      particle_num: 10000
      use_occlusion_likelihood: true
      use_range_likelihood: true
      range_likelihood_local_min_z: 0.05
      range_likelihood_local_max_z: 0.1
    </rosparam>
  </node>
  <node pkg="jsk_pcl_ros" type="extract_cuboid_particles_top_n" name="top_n_particles">
    <remap from="~input" to="plane_supported_cuboid_estimator/output/particles" />
  </node>
  <node pkg="nodelet" type="nodelet" name="top_n_particles_points" args="standalone pcl/ExtractIndices">
    <remap from="~input" to="plane_supported_cuboid_estimator/output/particles" />
    <remap from="~indices" to="top_n_particles/output" />
  </node>
  <node pkg="jsk_interactive_marker" type="dummy_camera.py" name="dummy_camera">
    <rosparam>
      frame_id: sensor_frame
    </rosparam>
  </node>
  <node pkg="jsk_pcl_ros" type="tf_transform_bounding_box_array" name="sensor_target">
    <remap from="~input" to="top_n_particles/output/box_array" />
    <rosparam>
      target_frame_id: sensor_frame
    </rosparam>
  </node>
  <node pkg="jsk_pcl_ros" type="tf_transform_bounding_box_array" name="sensor_candidate">
    <remap from="~input" to="sample_simulate_tabletop_cloud/output/candidate_boxes" />
    <rosparam>
      target_frame_id: sensor_frame
    </rosparam>
  </node>

  <node pkg="jsk_pcl_ros" type="boundingbox_occlusion_rejector" name="boundingbox_occlusion_rejector">
    <remap from="~input/camera_info" to="dummy_camera/camera_info" />
    <remap from="~input/target_boxes" to="sensor_target/output" />
    <remap from="~input/candidate_boxes" to="sensor_candidate/output" />
  </node>
  
  <group if="false">
  <node pkg="jsk_rqt_plugins" type="rqt_histogram_plot" name="x_histogram"
        args="/plane_supported_cuboid_estimator/output/histogram/global/x" />
  <node pkg="jsk_rqt_plugins" type="rqt_histogram_plot" name="y_histogram"
        args="/plane_supported_cuboid_estimator/output/histogram/global/y" />
  <node pkg="jsk_rqt_plugins" type="rqt_histogram_plot" name="z_histogram"
        args="/plane_supported_cuboid_estimator/output/histogram/global/z" />
  <node pkg="jsk_rqt_plugins" type="rqt_histogram_plot" name="roll_histogram"
        args="/plane_supported_cuboid_estimator/output/histogram/global/roll" />
  <node pkg="jsk_rqt_plugins" type="rqt_histogram_plot" name="pitch_histogram"
        args="/plane_supported_cuboid_estimator/output/histogram/global/pitch" />
  <node pkg="jsk_rqt_plugins" type="rqt_histogram_plot" name="yaw_histogram"
        args="/plane_supported_cuboid_estimator/output/histogram/global/yaw" />
  <node pkg="jsk_rqt_plugins" type="rqt_histogram_plot" name="dx_histogram"
        args="/plane_supported_cuboid_estimator/output/histogram/dx" />
  <node pkg="jsk_rqt_plugins" type="rqt_histogram_plot" name="dy_histogram"
        args="/plane_supported_cuboid_estimator/output/histogram/dy" />
  <node pkg="jsk_rqt_plugins" type="rqt_histogram_plot" name="dz_histogram"
        args="/plane_supported_cuboid_estimator/output/histogram/dz" />
  </group>
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find jsk_pcl_ros)/config/plane_supported_cuboid_estimator.rviz" />
  
</launch>
